File: apps/mobile/ios/RoamPulseWidget/RoamPulseWidget.swift
import SwiftUI
import WidgetKit

/// The "glance journey" from docs/PRODUCT_DISCOVERY.md §3 — the user
/// checks connectivity/data/expiry from the home screen without opening
/// the app. Structurally mirrors Android's RoamPulseWidget.kt: reads
/// whatever `WidgetDataStore` last persisted (written by AppDelegate's
/// MethodChannel handler) and renders it through the pure
/// `mapToWidgetUiState`; this file stays a thin composition shell with
/// no formatting logic of its own.
struct RoamPulseEntry: TimelineEntry {
    let date: Date
    let uiState: WidgetUiState
}

struct RoamPulseTimelineProvider: TimelineProvider {
    func placeholder(in context: Context) -> RoamPulseEntry {
        RoamPulseEntry(date: Date(), uiState: mapToWidgetUiState(emptyData()))
    }

    func getSnapshot(in context: Context, completion: @escaping (RoamPulseEntry) -> Void) {
        completion(currentEntry())
    }

    func getTimeline(in context: Context, completion: @escaping (Timeline<RoamPulseEntry>) -> Void) {
        // The widget only changes when Flutter pushes new data via the
        // MethodChannel (which calls WidgetCenter.reloadAllTimelines()) —
        // there's no periodic refresh schedule beyond that push.
        completion(Timeline(entries: [currentEntry()], policy: .never))
    }

    private func currentEntry() -> RoamPulseEntry {
        RoamPulseEntry(date: Date(), uiState: mapToWidgetUiState(WidgetDataStore().read()))
    }

    private func emptyData() -> WidgetPersistedData {
        WidgetPersistedData(
            connectionState: nil, carrierName: nil, technology: nil,
            connectivityLastSyncedAt: nil, destinationCity: nil, countryCode: nil,
            dataRemainingMb: nil, dataAllowanceMb: nil, daysRemaining: nil
        )
    }
}

struct RoamPulseWidgetEntryView: View {
    var entry: RoamPulseTimelineProvider.Entry

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(entry.uiState.connectionLabel)
                .font(.headline)
                .foregroundColor(toneColor(entry.uiState.connectionTone))
            if let carrierLine = entry.uiState.carrierLine {
                Text(carrierLine).font(.caption).foregroundColor(.secondary)
            }
            Spacer(minLength: 4)
            if let destination = entry.uiState.destinationLine {
                Text(destination).font(.subheadline).fontWeight(.medium)
            }
            if let remaining = entry.uiState.dataRemainingLabel {
                Text("\(remaining) remaining").font(.title3).fontWeight(.bold)
            }
            if let expiry = entry.uiState.expiryLabel {
                Text(expiry).font(.caption).foregroundColor(.secondary)
            }
            if let updated = entry.uiState.lastUpdatedLabel {
                Text("Updated \(updated)").font(.caption2).foregroundColor(.gray)
            }
        }
        .padding()
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
    }
}

private func toneColor(_ tone: ConnectionTone) -> Color {
    switch tone {
    case .positive: return Color(red: 0.18, green: 0.49, blue: 0.20)
    case .warning: return Color(red: 0.98, green: 0.66, blue: 0.15)
    case .negative: return Color(red: 0.78, green: 0.16, blue: 0.16)
    case .neutral: return Color(red: 0.38, green: 0.38, blue: 0.38)
    }
}

struct RoamPulseWidget: Widget {
    let kind: String = "RoamPulseWidget"

    var body: some WidgetConfiguration {
        StaticConfiguration(kind: kind, provider: RoamPulseTimelineProvider()) { entry in
            RoamPulseWidgetEntryView(entry: entry)
        }
        .configurationDisplayName("RoamPulse")
        .description("Shows your connection status and remaining data at a glance.")
        .supportedFamilies([.systemSmall, .systemMedium])
    }
}

@main
struct RoamPulseWidgetBundle: WidgetBundle {
    var body: some Widget {
        RoamPulseWidget()
    }
}

File: apps/mobile/android/app/src/main/kotlin/com/example/roam_pulse/widget/RoamPulseWidget.kt
package com.example.roam_pulse.widget

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.glance.GlanceId
import androidx.glance.GlanceModifier
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.provideContent
import androidx.glance.background
import androidx.glance.layout.Column
import androidx.glance.layout.Spacer
import androidx.glance.layout.fillMaxSize
import androidx.glance.layout.height
import androidx.glance.layout.padding
import androidx.glance.text.FontWeight
import androidx.glance.text.Text
import androidx.glance.text.TextStyle
import androidx.glance.unit.ColorProvider

/**
 * The "glance journey" from docs/PRODUCT_DISCOVERY.md §3 — the user
 * checks connectivity/data/expiry from the home screen without opening
 * the app. Reads whatever [WidgetDataRepository] last persisted (written
 * by MainActivity's MethodChannel handler) and renders it through the
 * pure [mapToWidgetUiState]; this class itself stays a thin composition
 * shell with no formatting logic of its own.
 */
class RoamPulseWidget : GlanceAppWidget() {
    override suspend fun provideGlance(context: Context, id: GlanceId) {
        val uiState = mapToWidgetUiState(WidgetDataRepository(context).read())
        provideContent { WidgetContent(uiState) }
    }
}

@Composable
private fun WidgetContent(state: WidgetUiState) {
    Column(
        modifier = GlanceModifier
            .fillMaxSize()
            .background(Color.White)
            .padding(12.dp),
    ) {
        Text(
            text = state.connectionLabel,
            style = TextStyle(
                fontSize = 15.sp,
                fontWeight = FontWeight.Bold,
                color = ColorProvider(toneColor(state.connectionTone)),
            ),
        )
        state.carrierLine?.let {
            Text(text = it, style = TextStyle(fontSize = 12.sp, color = ColorProvider(Color.DarkGray)))
        }
        Spacer(modifier = GlanceModifier.height(8.dp))
        state.destinationLine?.let {
            Text(text = it, style = TextStyle(fontSize = 13.sp, fontWeight = FontWeight.Medium))
        }
        state.dataRemainingLabel?.let {
            Text(
                text = "$it remaining",
                style = TextStyle(fontSize = 20.sp, fontWeight = FontWeight.Bold),
            )
        }
        state.expiryLabel?.let {
            Text(text = it, style = TextStyle(fontSize = 12.sp, color = ColorProvider(Color.DarkGray)))
        }
        state.lastUpdatedLabel?.let {
            Text(
                text = "Updated $it",
                style = TextStyle(fontSize = 10.sp, color = ColorProvider(Color.Gray)),
            )
        }
    }
}

private fun toneColor(tone: ConnectionTone): Color = when (tone) {
    ConnectionTone.POSITIVE -> Color(0xFF2E7D32)
    ConnectionTone.WARNING -> Color(0xFFF9A825)
    ConnectionTone.NEGATIVE -> Color(0xFFC62828)
    ConnectionTone.NEUTRAL -> Color(0xFF616161)
}
